### PR TITLE
DAOS-6722 placement: placement algorithm improvements

### PR DESCRIPTION
**Pull request by dmiter**
The issue with current implementation is it spend a lot of time to
find unused target when try to widely place. It calculate new hash
in the loop until it will point to next unused target. The time of
this calculation is unpredictable and in worst case can take a very
long time.

The idea of improvement is to make the time of placement predictable
and independent from widely placing. Now if first hash points to
used target it try to find first unused target around proposed in
the window of 64 targets size. If it still fails (all 64 targets
are used in this window) it search unused target in whole available
targets array. So, only one hash calculation is required now and in
worst case two linear searches into bits arrays.

Additionally it searches for root domain only once and then pass it
into subsequent functions.

Same of structure members were rearranged for better cache utilization
and prefetching.

The debug dumps functions are called only if appropriate debug level
is set.

### CODE REVIEW

**Comment by dmiter** [comment]
Any comments? Do you need any clarification?

File: src/placement/jump_map.c [hunk 1]
```diff
@@ -307,6 +310,103 @@ get_num_domains(struct pool_domain *curr_dom, enum PL_OP_TYPE op_type)
...
+/*
+ * Select the bit position (from the most-significant bit)
+ *
+ * The following 64-bit code selects the position of the rth 1 bit when
+ * counting from the left.
+ *
+ * \param[in]   v       value to find position with
+ * \param[in]   r       bit's desired count [1-64].
+ *
+ * \return      The position of the rth 1 bit in v [0-63]
```
Comment (ryon-jensen) [inline src/placement/jump_map.c:323]: Curious what the behavior is when the requested rth bit is higher than total 1 bits. For example, the rth 1 bit is requested, but there are r-1 1 bits? Does it just return 0? 0 is a valid position. Not sure it even matters for the usage of the function.

**Review by ryon-jensen — COMMENTED** [review]
Does this PR need to include the Features pragma for additional testing? Modifying the placement algorithm should probably include Rebuild and OSA tests, unless those are part of the standard PR testing now. I'm not very good at keeping track.

File: src/placement/jump_map.c (hunk 1 above)
Comment (dmiter) [reply src/placement/jump_map.c:323]: r is always 1 in this usecase. This means give me just first unused bit (first bit set to 1). If all bits are zero it returns 0. Then I check it again if this target is used or not.

**Comment by byron-marohn** [comment]
I think we need to have a meeting for you to explain to us how this works. That timezone are you in? Maybe we can set something up for this week?

My concern is that the jump-retry algorithm is not a side-effect of placing widely striped objects - it is actually the core of the algorithm. There are many corner cases it was designed to handle to ensure that data availability is always maintained - especially during reintegration and addition. Test coverage is almost entirely disabled currently for these cases - so a clean CI run is not indicative of it actually working (unfortunately). 

I haven't studied how this works closely, but one of the hardest use cases we had to design around was: 
- You place an object that spans 80% of targets. Let's say you have 10 ranks & 8 targets each, and you place something that's 2G32 (64 targets). 
- The data will be split into 32 distinct data segments, each with two copies. Each of those two copies must be placed on a different node, so that any single node failure will be recoverable. 
- You have one device fail, so the ~7 elements that are on the 8 targets that failed must be rebuilt elsewhere, and they themselves must be placed so that they again aren't on the same node as their replica. 
- You have another device fail, so its elements again need to be rebuilt elsewhere. The current algorithm I believe doesn't guarantee redundancy requirements are met in this multi-failure scenario, but it does ensure that the data is still readable / correctly rebuilt. 
- New capacity is added, migrating some of the original placements that were on failed targets to new capacity, and placing some data that wasn't failed onto failed devices, which then has to be migrated to fallback targets. 
- The first failure gets reintegrated. This has to move data back, while still ensuring that the placement is stable during the reintegration, so that active I/O can still proceed. 

None of this is fundamentally opposed to the description of the new approach you included here - but I want to make sure it really does what we think it needs to do. In particular, I'm concerned that batching things in this way will make it hard to ensure that for all object classes that the redundancy requirements are met. This is easy to reason about in the current scheme - because it operates the same brute-force way for each additional shard, but hard to reason about when you start adding fixed-sized buckets (collections of 64 targets, for example).

**Comment by byron-marohn** [comment]
Would also be helpful to have some idea how much this improved performance (probably a lot?). 

I'd also like to get @johannlombardi and @gnailzenh's input on whether we want to try to get this into 1.2. I think there's a pretty significant risk involved with introducing a big change like this at this point in the process.

**Comment by dmiter** [comment]
Let's arrange the meeting and discus this. I'm in UTC/GMT +3 hours timezone. Usually I'm working till very late. It's fine for me to use 7am-9am CDT (Central Daylight Time) timeslot.

Unfortunately I'm not familiar how the placement algorithm works in details. But looking into what this function is doing I just re-implemented it in different way. I didn't change any restriction to selected target. I just change how it's selected from array of available targets. My understanding is the logic you describe is above this function. The goal of this function is just return pseudo random target from an array of available targets.

The sequence of returned targets is changed. But it should be consistent if all input parameters are the same.

**Comment by jolivier23** [comment]
> Let's arrange the meeting and discus this. I'm in UTC/GMT +3 hours timezone. Usually I'm working till very late. It's fine for me to use 7am-2pm CDT (UTC/GMT -6 hours) timeslot.
> 
> 
> 
> Unfortunately I'm not familiar how the placement algorithm works in details. But looking into what this function is doing I just re-implemented it in different way. I didn't change any restriction to selected target. I just change how it's selected from array of available targets. My understanding is the logic you describe is above this function. The goal of this function is just return pseudo random target from an array of available targets.
> 
> 
> 
> The sequence of returned targets is changed. But it should be consistent if all input parameters are the same.

So my question in the recent call was whether the targets are the same as today.  If we make this change, will you be able to open a pool created from previous algorithm and find the data at the same place or does it change the expected location.  If it's consistent with prior version, it's less risky.  If not, it's a format change and we would absolutely want to make it sooner if we do it at all.

**Comment by dmiter** [comment]
Placement layout will be different with this patch. Therefore old formating is not supported. If we have correct versioning support for those layouts we can try to support both formats.

**Comment by dmiter** [comment]
Performance improvement of this patch is following. The values are the count of placement per second.

num shards | orig placement | new placement
8 | 538177 | 624403
16 | 289017 | 338378
32 | 152361 | 174655
64 | 77462 | 87565
128 | 38209 | 44450
256 | 18965 | 22001
512 | 8566 | 10905
1024 | 1578 | 5062
2048 | 802 | 2468
4096 | 399 | 1250
8192 | 196 | 606
15564 | 93 | 305

**Comment by jolivier23** [comment]
> Placement layout will be different with this patch. Therefore old formating is not supported. If we have correct versioning support for those layouts we can try to support both formats.

Looks like there are git conflicts now.  Do we have any means of detecting this layout change so we can report it?

**Comment by dmiter** [comment]
> Looks like there are git conflicts now. Do we have any means of detecting this layout change so we can report it?

Not yet. I just implement new version of placement for review. So, if we decide this can be used I can work for layout version detection and probably supporting both.

File: src/placement/jump_map.c (hunk 1 above)
Comment (dmiter) [reply src/placement/jump_map.c:323]: Simple linear search to process 2^32 searches for 227.46 sec. This function will do the same for 9.41 sec.

**Review by jolivier23 — CHANGES_REQUESTED** [review]
If you plan to continue with this patch (I know it's been discussed but I'm not sure where we landed), it will need conflict resolution.

**Pull request closed without merging**